src/Microsoft.FeatureManagement/FeatureManagerSnapshot.cs

```diff
@@ -13,8 +13,8 @@ namespace Microsoft.FeatureManagement
     class FeatureManagerSnapshot : IFeatureManagerSnapshot
     {
         private readonly IFeatureManager _featureManager;
-        private readonly IDictionary<string, bool> _flagCache = new Dictionary<string, bool>();
-        private IEnumerable<string> _featureNames;
+        private readonly Dictionary<string, Task<bool>> _flagCache = new Dictionary<string, Task<bool>>();
```

Comment:
`ConcurrentDictionary<K,V>` would be possible, but the factory in `GetOrAdd` becomes the overload with state for .NET 5, which isn't used here. So in order to avoid a closure (= allocation) a plain `Dictionary<K,V>` is used, that is locked only in the case when no entry is found (so on the cold-path).

Reply:
I would still use `ConcurrentDictionary<K,V>` here to resolve any race conditions. They are much better at fine-grain locking than what we could do here in a simple manner. Also, introducing any lock management will automatically make this code more complex to maintain. `ConcurrentDictionary` should provide us the simplest fix, although it may not have the optimization that you mentioned.

Reply:
Will update to the CDN.
Just re-checked and the optimization I meant got added to .NET Core 2.0 and .NET Standard 2.1, so it's usable here 😃